### ISSUE TITLE
Names module DNS getHostByName error on certain sites

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3998]
|<img alt="pythonologist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pythonologist reported|
|-|-|
|Trac ID|trac#3998|
|Type|release blocker: regression|
|Created|2009-09-03 17:41:02Z|
|Branch|https://github.com/twisted/twisted/tree/names-no-transport-3998|

With the attached program I get this output:

```
D:\software\Twisted-trunk\twisted\names\dns.py:1662: DeprecationWarning: Please only pass IPs to write(), not hostnames
  self.transport.write(message.toStr(), address)
result: [Failure instance: Traceback: <type 'exceptions.AttributeError'>: 'NoneType' object has no attribute 'stopListening'
D:\software\Twisted-trunk\twisted\internet\udp.py:126:doRead
D:\software\Twisted-trunk\twisted\names\dns.py:1693:datagramReceived
D:\software\Twisted-trunk\twisted\internet\defer.py:238:callback
D:\software\Twisted-trunk\twisted\internet\defer.py:307:_startRunCallbacks
--- <exception caught here> ---
D:\software\Twisted-trunk\twisted\internet\defer.py:323:_runCallbacks
D:\software\Twisted-trunk\twisted\names\common.py:206:<lambda>
]
```

Certain domain names appear to return a particular set of DNS records requiring recursion, which does not appear to work properly: one example is www.bbc.co.uk - note that plain bbc.co.uk works fine. Possibly related to [#2850](https://github.com/twisted/twisted/issues/2850).

Attachments:

* [dns_test_2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/adb/adb73dae576eeb8e2819baa713cd8d99d203b140/bca0ebe560ddff58c3436c6aede17d15740cd1fc.py) (396 bytes) - added by pythonologist on 2009-09-03 17:42:10Z - simple dns test program
* [gethostbyname-dns.pcap](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/adb/adb73dae576eeb8e2819baa713cd8d99d203b140/807fbb19bfe309a17fadbfc2b4fe9d6ca82872c0.pcap) (738 bytes) - added by exarkun on 2009-09-16 03:05:18Z - 
* [3998.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/adb/adb73dae576eeb8e2819baa713cd8d99d203b140/cf41cc0f77d2d9a043eb117e6f9d67bb371f0e4a.diff) (842 bytes) - added by smcq on 2009-09-16 23:13:40Z - Patch that fixes problem.  extractRecord incorrectly attempted to close it's resolver's connection.

<details><summary>Searchable metadata</summary>

```
trac-id__3998 3998
type__release_blocker__regression release blocker: regression
reporter__pythonologist pythonologist
priority__normal normal
milestone__ 
branch__branches_names_no_transport_3998 branches/names-no-transport-3998
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1251999662000000 1251999662000000
changetime__1284216517000000 1284216517000000
version__None None
owner__ 
cc__smcq cc__exarkun cc__jesstess cc__Scramblejams
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to pythonologist|
|-|-|

I cannot reproduce this.  Please attach a DNS traffic capture.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

smcq ran into this error, apparently, and brought it up on IRC.  On taking a second look, I was able to reproduce this failure after changing dns_test_2.py to use `twistedmatrix.com` instead of `www.bbc.co.uk`.  This difference is probably just due to differences in the networks pythonologist and I were running the example on, but I haven't dug into the problem yet so that's just a guess.

I am attaching a pcap file of the dns traffic generated by the script when I run it.  Examination (or use, in the worst case) of this should be enough to reproduce the issue in a test case.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

A guess I might make is that [#3342](https://github.com/twisted/twisted/issues/3342) introduced this bug (that would make this a regression :/).

**Comment by twisted-trac** [comment]
|<img alt="smcq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|smcq set owner to smcq<br>smcq set status to `assigned`|
|-|-|

I have tracked this bug to the NS followup query done by extractRecord in common.py.

return r.lookupAddress(str(name)
                ).addCallback(lambda (ans, auth, add): extractRecord(r, name, ans + auth + add, level - 1)
                ).addBoth(lambda passthrough: (r.protocol.transport.stopListening(), passthrough)[1])

Simply removing the addBoth (which does appear entirely unnecessary by my reading) will fix the bug.

To duplicate, use a NS record, I was able to successfully duplicate it with saucelabs.com.  You may need to set retries to two in order to cause the bug (unsure as to reason, and you may want to dig into that before patching).

With the addBoth removed the following commands successfully run, with it enabled the following commands will cause the server to crash with NoneType not having stopListening.  

If you run the following you should be able to reproduce the error against a trivial nameserver implementation: 

dig +tries=2 @localhost -p 4053 saucelabs.com

**Comment by twisted-trac** [comment]
|<img alt="smcq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|smcq commented|
|-|-|

Sorry, I didn't mean to accept this, however I can submit a patch in the morning for my resolution.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess set status to `new`|
|-|-|

Thanks for the patch smcq!

Can you submit a test case as part of this patch? Also, once you're ready for someone to review a patch, please add 'review' to the keywords property and reassign to nobody so it doesn't get lost in the shuffle.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

Oh trac thou dost protesteth too much

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

[#4210](https://github.com/twisted/twisted/issues/4210) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml commented|
|-|-|

Twisted 10.1.0 has been released. Removing this ticket from the 10.1 milestone. If you think it's severe enough to demand a follow-up release, please say so and add this ticket back to the 10.1 milestone.

**Comment by twisted-trac** [comment]
|<img alt="Scramblejams's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|Scramblejams commented|
|-|-|

This was a blocker for me, now I have to manually patch all my Twisted installations. And patch them again whenever a security update comes out... Are we just waiting on a test case?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [29989]) Branching to 'names-no-transport-3998'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner|
|-|-|

OK, I did the bare minimum, adding a test failing before and not after. Considering the fix is obvious, I'll it's good enough.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml set owner to @therve|
|-|-|

Thanks for adding the test therve.

I'm a little confused though as to why the test builds a `messages` list only to have it returned from `query`. Wouldn't it be simpler to have `query` define and return `m`?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml commented|
|-|-|

```
<therve> jml, re your comment, I want it to blow up if another query is made
<therve> I guess it doesn't matter that much if the test is passing
<jml> there are eight tickets marked to review, hard to say if they've got patches or not.
<jml> therve, maybe add a comment saying "Blow up if another query is made", as well as saying why you want to do that.
<therve> jml, sounds good
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve removed owner|
|-|-|

Thanks, I added a comment in r29999.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_18"></a>|@jml set owner to @therve|
|-|-|

Thanks. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set status to `closed`|
|-|-|

(In [30001]) Merge names-no-transport-3998

Author: therve
Reviewer: jml
Fixes: [#3998](https://github.com/twisted/twisted/issues/3998)

Fix a regression twisted.names.common.extractRecords where it tried to close a
non-present transport.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
